### DISCUSSION TITLE
How can I replace values with predefined variables in .egsinp files?

### DISCUSSION THREAD

**Discussion by Sasa-fras** [nrc-cnrc/EGSnrc#1099] [Questions]
I'm curious if there is any way to replace the values like the ones that show up when defining geometry with variables that I can define all in one spot somewhere else in the .egsinp file. Since many of the values in my geometry definition are relative to the lengths of other geometric objects it would be much more convenient if I could replace x values of planes with something like 'x_1' and somewhere else define 'x_1 = 2.70 + x_0' and such so that I can change the thickness of something in one spot without having to go adjust values for 8 different spots. I tried looking around already and found something like :start key: which allowed me to make 'keys' and assign them values (basically defining variables except they all need to be named in a specific way, I think), but that didn't work at all and the geometry threw me an error that was basically saying my input is not accepted. I saw that a contributor suggested adding this feature in 2017, has it ever been added? (that was issue #697)

**Comment by rtownson** [comment]
The closest to this that we have in egs++ so far are `input loops`, which you can find documentation for on this page:
https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/common.html

Personally, I just use a python script to do this. Instead of a .egsinp file I create a file that's a template, e.g. a .template file. Inside I use python keys for variable replacement (e.g. %(myVariableName)s would be for a string replacement), then write out the result to .egsinp file(s). This allows me to create many .egsinp files where one or several parameters are changing. 

Something like this:
```
content = template.read() % variables
```
where `variables = vars(self)`, does the replacement from python.

**Reply by Sasa-fras** [reply]
Thanks for the response. It sounds like the method using python would be the best for me. Do you have any example files I could look at so that I can try to replicate the method you are describing. I think I understand what you're saying that you do but I don't know how to implement it.

**Reply by rtownson** [reply]
Please send me an email (found by the link below), and I can share a script to get you started.

https://nrc.canada.ca/en/research-development/products-services/software-applications/egsnrc-software-tool-model-radiation-transport

**Reply by ftessier** [reply]
Indeed, input loops are great, but there is no facility to do simple math with the loop variables, scripting is best at that point, because you can then implement any calculation or logic with input parameters. @rtownson, feel free to share our `egs-rollout` script with @Sasa-fras, I hope we can include it in the release this year!
